README.md

```diff
@@ -1 +1,189 @@
+>
+> — [RFC8259](https://www.rfc-editor.org/rfc/rfc8259#page-10)
+
+``` json
+{
+  "role": "user",  // Parsed by an IPLD implementation
+  "role": "admin", // Malicious duplicate field, omitted by the IPLD parser, accpeted by the browser
+  "links": [
+    {"/": "bafkreidb2q3ktgtlm5yio7buj3sypyghjtfh5ernsteqmakf4p2c5bwmyi"},
+    {"/": "bafkreic75ydg5vkw324oqkcmqltfvc3kivyngqkibjoysdwiilakh4z5fe"},
+    {"/": "bafkreiffdiz6raf46zrr3b2usufgz5fo44aggmocz4zappr6khhhljcdpy"}
+  ],
+  "sig": "8ufaS9w3CGN8cbQTUSoL1i7eaKiWLSXsD2LbZVmvM9zF"
+}
+```
+
+In the above example, the canonicalization step MAY lead to the signature validating, but the client parsing the `role: "admin"` field instead.
+
+# 2 Safety
+
+Data that has already been parsed to an in-memory IPLD representation can be canonically encoded to a trivially: it has already been through a [parser / validator](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/).
+
+Data purporting to conform to an IPLD encoding (such as [DAG-JSON](https://ipld.io/specs/codecs/dag-json/spec/)) MUST be validated prior to signature verification. This can be as simple as round-trip decoding/encoding the JSON and checking that the hash matches. A validation error MUST be signalled if it does not match. 
+
+> [Implementers] may provide an opt-in for systems where round-trip determinism is a desireable [sic] feature and backward compatibility with old, non-strict data is unnecessary.
+>
+> — [DAG-JSON Spec](https://ipld.io/specs/codecs/dag-json/spec/)
+
+As it is critical for signatures guard against various attacks, the assumptions around canonical encoding MUST be enforced.
+
+## 2.1 Signing CIDs
+
+Rather than validating the inline IPLD, replacing the data with a CID link to the content MAY be used instead. Note while this is very safe (as it impractical to alter a signed hash), this approach mixes data layout with security, and may have a performance, disk, and networking impacts.
+
+## 2.2 Raw (Noncanonicalized) Data
+
+Canonicalization is not required if data is encoded as raw bytes (multicodec `0x00`). The exact bytes are already present, and MUST not be changed.
+
+# 3 Varsig Format
+
+A varsig is a bytestring that includes the following information:
+
+| Segment Name              | Type     | Description                                     | Required |
+|---------------------------|----------|-------------------------------------------------|----------|
+| Varsig Prefix             | `0xd000` | The multicodec varsig prefix                    | Yes      |
+| Public Key Prefix         | `Varint` | The multicodec prefix for the public key type   | Yes      |
```

Comment:
Not disagreeing, but is there reason why above two can't be one ? We have already registered codes for relevant keys in the multicodec table

Reply:
They can, but we need to register new signature types as new keys get added. It's not the end of the world, but would be nice to not need to go back into the spec every time.

Reply:
I agree that it's annoying that we've already registered those codes. Am I right to guess that despite them being labelled as `draft` that they're actually quite difficult to remove?

Reply:
Perhaps I should explain some of my reasoning. There's two ways of looking at the design space, both of which are compatible with the overall approach in this spec.

# 1. Signatures are Only Cryptographic

In this worldview, there is one and only one algorithm with which to sign EdDSA. In fact, there signature types are in a one-to-one correspondence with public key types (AFAIK at least. IANAC (I am not a cryptographer)).

# 2. Signatures Include Prefixes

I think this is where we started with this spec, but the line is really blurry. Prefixing the magic bytes from EIP191 before hashing & signing is more like part of the hash step, but you could call that entire process "the signature algorithm" (they are sometimes classified as such — `RS256` is RSA + SHA256).

I think that working through this spec in detail has started to make me think that decoupling the encoding from the hash from the signature algorithm is a good idea. It's much more composable, put a clear boundary on where the encoding begins and the signature ends. It makes "signing arbitrary payloads with Passkey" impossible unless we want to specify a multicodec for passkey, but we can also say that this method is "nonstandard" (it is!), and the application needs to understand more about the payload to do the encoding. It's sure inconvenient, but I don't currently see a clean way to handle every possible case (though I could be overlooking something!), especially when we get into them being nested (ReCap inside a JWS inside an EIP191).

Reply:
I understand reasoning for breaking key type, hash type and format into their own pieces however that does not inherently requires separate tag to mark all varsigs, same just as well could be achieved using e.g. key types. They will have to be distinct from the actual key codecs in mulitcode table, but my point is we don't need another wrapper first tag can imply semantics.

Only rationale I can see to (lets call it a) namespace tag is to reuse same key codes, but I fail to recognize benefits of it. Furthermore that is not the direction that [multihash](https://multiformats.io/multihash/) chose, they do not have multihash namespace tag instead hash tag signals that it is a multihash.

So unless we have a good reason to go different direction and pay bit of size overhead, I'd argue we should drop the namespace tag and just use the codes that we've already allocated for varsig as varsig+key signalling mechanism.

Reply:
> Furthermore that is not the direction that [multihash](https://multiformats.io/multihash/) chose

@Gozala Good point!

> So unless we have a good reason to go different direction and pay bit of size overhead, I'd argue we should drop the namespace tag and just use the codes that we've already allocated for varsig as varsig+key signalling mechanism.

The size overhead is one byte (since changing the proposal to use `0x34`), which is the same amount of byte overhead that's in the current signature codes. But I'm comfortable with this ☝️ @oed did raise a good point that we're going to be registering a LOT of signature types for every variation.

Having thought about it a bit more, I may have a reason why we might actually want canonicalization info as multihashes instead of special varsig codes: domain separation also falls into this class of payload transformation. First some background:

If you have an application that needs two hash functions (e.g. to increase accuracy in an invertible bloom filter), it is completely legitimate to create a "new hash" by appending some constant bytes, and having that be `hash2`. This puts the output in a completely different part of the hash space, and is thus "a different hash function".

```js
const hash1 = (payload) => sha256(payload)
const hash2 = (payload) => sha256(payload + "example.com")
```

HKDF bakes the above idea right into the hashing algorithm: you can supply it an info tag used for domain separation.

EIP-191 is a form of domain separation. You prepend some the bytes that say something like `"This is an Ethereum transaction: "` to the front of the message before running it through the standard hash function, ahead of signing it.

Arguably we could register some multihashes that are `EIP191`, `EIP712`, etc. Changing multihash itself to accept domain separation params seems like a high bar, but could conceivably be done for e.g. HKDF.

---

Anyhow, I'm also comfortable with registering a bunch of varsig codes. It would be the exact same size as we have in the current proposal, and we'd have to register signature types more often, but it's also _fine_. I'll let both of you weigh in, though.

Reply:
Just to be clear I’m not opposing umbrella tag, I just want to make sure it’s deliberate choice.

> and we'd have to register signature types more often, but it's also fine.

I’m not sure I follow this, I would imagine only codes needed would correspond to public key types, hashing and other details follow in their own tags don’t they ?

I’m also not sure one could get away without registrations, it may be in a separate table but registration in one never the less. I could be missing something though

Reply:
Perhaps to clarify the reasoning, though I am known to go too granular: signature algorithms like [ECDSA](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf) don't force a _particular_ hashing algorithm.

> for some signatures, you can sort of determine the hash used by using the key. With RSA (but apparently not DSA or ECDSA [...]) you can (in some cases) use the public key to revert a signature to the cryptographic digest from which it was created
> 
> — https://security.stackexchange.com/questions/174213/which-hash-algorithm-to-use-for-ecdsa-signature-verification-openssl

So handling all of this at the level of signature codes means encoding the $n\times{m}$ of key type, hash function, sometimes length, and possibly other factors like the fact that RSA-PSS adds a (nondeterministic) random nonce every time.

This RSA thing alone makes me agree with @Gozala's earlier comment: we need to bite the bullet of the $m\times{n}$ complexity to gain the flexibility of all of the various signature types for e.g. the many variations of RSA (and to a lesser extent the ECDSAs).

Thoughts?

Reply:
> I just want to make sure it’s deliberate choice

Also FWIW @Gozala — I super appreciate this! I often find myself in that somewhat thankless role. Thank you 🚀✨